transaction.go

```diff
@@ -153,49 +289,121 @@ func StartFunc(service, user string, handler func(Style, string) (string, error)
+// is not!) and using a LockOSThread/UnlockOSThread pair.
 func StartConfDir(service, user string, handler ConversationHandler, confDir string) (*Transaction, error) {
 	if !CheckPamHasStartConfdir() {
-		return nil, errors.New("StartConfDir() was used, but the pam version on the system is not recent enough")
+		return nil, NewTransactionError(ErrSystem,
+			errors.New("StartConfDir() was used, but the pam version on the system is not recent enough"))
 	}
 
 	return start(service, user, handler, confDir)
 }
 
+type panicHandlerFunc func(any)
+
+var defaultPanicHandler = func(v any) { panic(v) }
+
+var panicHandler panicHandlerFunc = defaultPanicHandler
+
 func start(service, user string, handler ConversationHandler, confDir string) (*Transaction, error) {
 	switch handler.(type) {
 	case BinaryConversationHandler:
 		if !CheckPamHasBinaryProtocol() {
-			return nil, errors.New("BinaryConversationHandler() was used, but it is not supported by this platform")
+			return nil, NewTransactionError(ErrSystem,
+				errors.New("BinaryConversationHandler() was used, but it is not supported by this platform"))
 		}
 	}
 	t := &Transaction{
 		conv: &C.struct_pam_conv{},
 		c:    cgo.NewHandle(handler),
 	}
+
+	callers := make([]uintptr, 10)
+	haveCallerInfo := runtime.Callers(2, callers) > 0
+	runtime.SetFinalizer(t, func(t *Transaction) {
+		handle := atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&t.handle)))
+		if handle == nil {
+			return
+		}
+		if !haveCallerInfo {
+			panicHandler(fmt.Sprintf("Transaction %p was never ended", t))
+			return
+		}
+
+		stackTrace := make([]string, 0, len(callers))
+		for range callers {
+			frame, more := runtime.CallersFrames(callers).Next()
+			stackTrace = append(stackTrace, fmt.Sprintf("%s:%d",
+				frame.File, frame.Line))
+			if !more {
+				break
+			}
+			callers = callers[1:]
+		}
+		panicHandler(fmt.Sprintf("Transaction %p was never ended. "+
+			"Initialization was at:\n %s",
+			t, strings.Join(stackTrace, "\n ")))
+	})
+
 	C.init_pam_conv(t.conv, C.uintptr_t(t.c))
-	runtime.SetFinalizer(t, transactionFinalizer)
 	s := C.CString(service)
 	defer C.free(unsafe.Pointer(s))
 	var u *C.char
 	if len(user) != 0 {
 		u = C.CString(user)
 		defer C.free(unsafe.Pointer(u))
 	}
+	var err error
 	if confDir == "" {
-		t.status = C.pam_start(s, u, t.conv, &t.handle)
+		err = t.handlePamStatus(C.pam_start(s, u, t.conv, &t.handle))
 	} else {
 		c := C.CString(confDir)
 		defer C.free(unsafe.Pointer(c))
-		t.status = C.pam_start_confdir(s, u, t.conv, c, &t.handle)
+		err = t.handlePamStatus(C.pam_start_confdir(s, u, t.conv, c, &t.handle))
 	}
-	if t.status != C.PAM_SUCCESS {
-		return nil, t
+	if err != nil {
+		var _ = t.End()
+		return nil, err
 	}
 	return t, nil
 }
 
-func (t *Transaction) Error() string {
-	return C.GoString(C.pam_strerror(t.handle, C.int(t.status)))
+// TransactionError is an interface that may be implemented to provide a more
```

Comment:
I like the idea, but don't you think a wrapped error would make more sense? Instead of adding a new type you could do something like:

```go
return fmt.Errorf("Something bad happened: %w", err)
```

Reply:
So... Not much in this branch but in the changes in #13 I do require to have an error I can inspect for its actual error code so that we can pass it back to C.

Now, unwrapping could help with that, but not sure if this makes things more problematic. I can try to see if I can handle both needs, but ideally having a specific error type is something that looked better.

Reply:
A part from that, the fact of being able to check for an error type in tests is quite convenient. Indeed this can be done with `error.is` + the `pam.Error` provided, and unwrap it too, but I'm wondering how convenient is to have on the user-side compared to have some utility struct that handles explicitly.

Reply:
I'm trying to change this so that I can avoid the `TransactionError` at all, it seems to go fine so far.

Reply:
Ok, removed... At the end, wrapping works well enough for all the purposes.